### PR TITLE
CreateSampleWorkspace to also allow creating monitors

### LINKED ISSUE #17477: CreateSampleWorkspace to allow adding monitors to the workspace
[CreateSampleWorkspace](http://docs.mantidproject.org/nightly/algorithms/CreateSampleWorkspace-v1.html) allows the creation of a workspace, with an instrument, for ease of testing (unit and manual). For ticket #17138 it would be useful if it could also create monitors.

This ticket will add an option to specify a number of monitor spectra, containg the same spectra as the detectors.

### PR DESCRIPTION
**Pull request by ianbush**
This adds the option "NumMonitors" to CreateSampleWorkspace.

**To test:**
- Run CreateSampleWorkspace with the NumMonitors and any a variety of other options, checking the monitors are added to the workspace (their workspace indices will be the after the detectors)
- Check the monitors are visible in the instrument view
- Check the updated documentation looks good

Fixes #17477 .

_Does not need to be in the release notes._

---
#### Reviewer

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):
##### Code Review
- [x] Is the code of an acceptable quality?
- [x] Does the code conform to the [coding standards](http://www.mantidproject.org/Coding_Standards)? Is it well structured with small focussed classes/methods/functions?
- [x] Are there unit/system tests in place? Are the unit tests small and test the a class in isolation?
- [ ] If there are changes in the release notes then do they describe the changes appropriately?
##### Functional Tests
- [x] Do changes function as described? Add comments below that describe the tests performed?
- [ ] How do the changes handle unexpected situations, e.g. bad input?
- [x] Has the relevant documentation been added/updated?
- [ ] Is user-facing documentation written in a user-friendly manner?
- [ ] Has developer documentation been updated if required?
- Does everything look good? Comment with the ship it emoji but don't merge. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: Framework/Algorithms/inc/MantidAlgorithms/CreateSampleWorkspace.h [hunk 1]
```diff
@@ -63,8 +63,9 @@ class DLLExport CreateSampleWorkspace : public API::Algorithm {
                            Geometry::Instrument_sptr inst,
                            const std::string &functionString, bool isRandom);
   Geometry::Instrument_sptr createTestInstrumentRectangular(
-      API::Progress &progress, int num_banks, int pixels, double pixelSpacing,
-      const double bankDistanceFromSample, const double sourceSampleDistance);
+      API::Progress &progress, int num_banks, int numMonitors, int pixels,
```
Comment (KarlPalmen) [inline Framework/Algorithms/inc/MantidAlgorithms/CreateSampleWorkspace.h]: ``numMonitors`` is not consistent with ``num_banks``.

File: Framework/Algorithms/src/CreateSampleWorkspace.cpp [hunk 2]
```diff
@@ -497,6 +504,35 @@ Instrument_sptr CreateSampleWorkspace::createTestInstrumentRectangular(
...
+       monitorNumber <= monitorsStart + numMonitors; monitorNumber++) {
+    // Make a new bank
+    std::ostringstream monitorName;
+    monitorName << "monitor" << monitorNumber;
+
+    RectangularDetector *bank = new RectangularDetector(monitorName.str());
+    bank->initialize(pixelShape, 1, 0.0, pixelSpacing, 1, 0.0, pixelSpacing,
+                     monitorNumber, true, 1);
+
+    // Mark them all as detectors
```
Comment (KarlPalmen) [inline Framework/Algorithms/src/CreateSampleWorkspace.cpp]: It looks like you are marking them all as monitors.

File: Framework/Algorithms/test/CreateSampleWorkspaceTest.h [hunk 3]
```diff
@@ -452,6 +452,63 @@ class CreateSampleWorkspaceTest : public CxxTest::TestSuite {
...
+    TS_ASSERT_DELTA(ews->readY(200)[50], 257, 0.0001);
+    TS_ASSERT_DELTA(ews->readY(200)[60], 7, 0.0001);
+
+    TS_ASSERT(ews->getDetector(201)->isMonitor());
+    TS_ASSERT_DELTA(ews->readY(201)[50], 257, 0.0001);
+    TS_ASSERT_DELTA(ews->readY(201)[60], 7, 0.0001);
+
+    // Remove workspace from the data service.
+    AnalysisDataService::Instance().remove("outWS");
+  }
```
Comment (KarlPalmen) [inline Framework/Algorithms/test/CreateSampleWorkspaceTest.h]: You set the number of monitors as 2, but test for 3 monitors. Perhaps, you need to test that the 3rd detector is not a monitor.

**Comment by KarlPalmen** [comment]
I found that the monitors were given large spectra numbers than the other detectors, which other workspaces I've seen with monitors have the smallest spectra numbers given to the monitors.

File: Framework/Algorithms/test/CreateSampleWorkspaceTest.h (hunk 3 above)
Comment (KarlPalmen) [inline Framework/Algorithms/test/CreateSampleWorkspaceTest.h]: Now I see that the first detector is tested to not be a monitor. I expected this to be the 3rd.

**Comment by ianbush** [comment]
I have now changed this so the monitors get the first spectrum numbers in the workspace. We have reversed this for some ILL instruments, but I accept this is not normal!

I have also fixed the comments and the inconsistent variable names.

**Comment by ianbush** [comment]
This is causing a crash in the SmoothNeighburs doc test, I will take a look as to why.

**Comment by KarlPalmen** [comment]
When looking at the instrument view, I found names like "Bank1", "Bank2", "Monitor3", "Monitor4". Would it be more appropriate to have something like "Bank1", "Bank2", "Montitor1", Monitor2"?

**Comment by ianbush** [comment]
Yes, I had not noticed that, it is now fixed so the monitors get the numbers you would expect.

**Comment by KarlPalmen** [comment]
:shipit:

**Review by martyngigg — APPROVED** [review]

**Pull request merged**